SoC and Java 8

jebl01 opened the pull request:
### What
* Updated to Java 8
* Split the code into more manageable and testable parts.
* Added more test
* Bumped JGroups dependency
* Updated doc
* Fixed the issue with "self filtering" not taking ports into account (and therefore trying to connect to itself)

### Why
To make testing and future code reasoning/refactoring easier.

ctrimble commented on pom.xml:
```diff
@@ -122,6 +126,16 @@
+        <artifactId>maven-compiler-plugin</artifactId>
+        <configuration>
+          <source>1.8</source>
+          <target>1.8</target>
```
Unless there is something in the parent pom to break this, we should just be setting:

```
    <maven.compiler.source>1.8</maven.compiler.source>
    <maven.compiler.target>1.8</maven.compiler.target>
```

jebl01 replied:
👍  NP! This was added by IntelliJ magic :-)

jebl01 replied:
You mean like this, right?
```xml
<java.source.version>1.8</java.source.version>
<java.target.version>1.8</java.target.version>
```

ctrimble replied:
Yes, use those.  I will come along later and move the project to a parent pom that is not redefining the standard properties.

ctrimble commented on src/main/java/com/meltmedia/jgroups/aws/AWS_PING.java:
```diff
@@ -127,8 +101,7 @@
  * <blockquote>
  * <pre>
- * &lt;com.meltmedia.jgroups.aws.AWS_PING
- *   timeout="3000"
```
thanks for cleaning up the docs.

ctrimble commented on pom.xml:
```diff
@@ -10,7 +10,7 @@
   <groupId>com.meltmedia.jgroups</groupId>
   <artifactId>jgroups-aws</artifactId>
-  <version>1.5.1-SNAPSHOT</version>
+  <version>2.0.0-SNAPSHOT</version>
```
If users can move from `1.5.0` to this version, without too much work, then we should probably just move to `1.6.0-SNAPSHOT`.

jebl01 replied:
Ok! Took a guess that the move from 1.7 to 1.8 was to consider breaking, but I can change!

ctrimble commented:
Thanks for this great PR, @jebl01 !  I put in a few minor notes, but overall this looks really good.  I think we should:

1. make the minor updates I pointed out, If you agree.
2. squash those pom changes in the `Added test dependency...` commit, keeping your overall commit structure.
3. I can merge these into develop and get a SNAPSHOT out there for testing.
4. when you are happy with the quality of that version, I will go ahead and release it.

Will that work for you?

ctrimble commented on src/test/java/com/meltmedia/jgroups/aws/WithEc2.java:
```diff
@@ -0,0 +1,26 @@
+import static org.mockito.Mockito.when;
+
+public interface WithEc2 {
+  default AmazonEC2 ec2Mock(Tag... tags) {
```
This seems like a strange use of default methods.  This should probably be a class called `Mocks` with a `public static AmazonEC2 ec2Mock(Tag... tags)` method and then statically import it into the test.

jebl01 replied:
Yes ;-) This is a habit (good or bad) that I have adopted for tests lately... guess it's my dream of mixins/traits that leaks. It becomes quite handy in larger projects, where you can have a lot of cross cutting test concerns... But If you think it's to wacky, I can always change :-)

ctrimble replied:
It just seems like you get the same thing from static test imports.  I can see how this might avoid large lists of static imports, when code formatters help out by expanding `*`.  You have taken your time to put your mark on this project, so I am not going to take a hard stance on this.  It is a style issue.  Just let me know when you think the code is ready to be merged.

jebl01 commented:
Hmm, guess that *squash / force push* wiped most of the PR comments we had going :P
Well, I fixed the things we were talking about:
* project version 2 -> 1.6
* Java 8 by overriding properties instead of introducing plugin
* Removed wacky "traits" thingie in favour of a `Mocks` class

ctrimble commented:
@jebl01 I pushed a 1.6.0-SNAPSHOT of these changes to Sonatypes OSS repository.  I also fixed problems with the parent pom and resolved some other build issues related to site generation.  Please verify that the version is working, with some practical test, and let me know if this is ready for release.

jebl01 commented:
Running 1.6.0-SNAPSHOT on AWS now, LGTM!

ctrimble commented:
@jebl01 This has been released.  Just waiting for Sonatype to sync it to Maven Central.  You can check http://repo1.maven.org/maven2/com/meltmedia/jgroups/jgroups-aws/ to see if it has been published yet.

jebl01 commented:
👍
